controller/uninstall_controller.go

```diff
@@ -315,13 +315,16 @@ func (c *UninstallController) uninstall() error {
 		}
 	}
 
-	// A race condition exists where manager may attempt to recreate certain
-	// CRDs after their deletion. We must delete manager and do a final check
-	// for CRDs, just in case.
+	// A race condition exists where manager may attempt to recreate certain CRs after their deletion, e.g. BackupTarget.
+	// We must delete manager first and then delete those CRs.
 	if waitForUpdate, err := c.deleteManager(); err != nil || waitForUpdate {
 		return err
 	}
 
+	if waitForUpdate, err := c.deleteRecreatedCRs(); err != nil || waitForUpdate {
+		return err
+	}
```

Comment:
We need to seperate BackupTarget CR deletion out of `deleteCRDs()`
because BackupTarget CR cannot be deleted before manager deletion
In this PR we move these kinds of CR deletion out of `deleteCRDs()`
So run `deleteCRDs()` won't work. because BackupTarget deletion is not there

On the other hand, move BackupTarget CR deletion back to `deleteCRDs` and run the `deleteCRDs()` again, that is the original implementation, and it won't work.

cc @innobead

Reply:
OK. Just discussed with @ChanYiLin, after `deleteCRDs()`issued a deletion of backuptarget CRs, `deleteCRDs()` will return and wait for the deletion. However, longhorn-manager might recreate a new backuptarget CR.  `deleteCRDs` issued a deletion again. This behavior will lead to timeout error.

The fix looks good to me. Only need to
- rename `deleteCRDs` to `deleteCRs()`
- leave a comment for discerning `CR` ad `RecreatedCR`

Reply:
The root cause of this issue
- In deleteCRDs we delete the BackupTarget if it is found
- We return and wait for the next reconcile to see if the resource is truly deleted by checking `len(BackupTarget) == 0`
- We only start to delete manager when we make sure all the CRs are deleted.
- But there is a large chance that the BackupTarget just got recreated between two reconciliation so the `len(BackupTarget) == 0` will never be true

Along this 5 years, the uninstallation succeeded only because BackupTarget was not recreated fast enough so the `len(BackupTarget) == 0` happened to be true